commons-numbers-complex-arrays/src/main/java/org/apache/commons/numbers/complex/arrays/ComplexList.java

```diff
@@ -0,0 +1,357 @@
+        return Complex.ofCartesian(realAndImagParts[i], realAndImagParts[i + 1]);
+    }
+
+    /**
+     * Replaces the element at the specified position in this list with the specified element's
+     * real and imaginary parts. No range checks are done.
+     *
+     * @param index - Index of the element to replace.
+     * @param real - Real part \( a \) of the complex number \( (a +ib) \).
+     * @param imaginary - Imaginary part \( b \) of the complex number \( (a +ib) \).
+     */
+    private void setNoRangeCheck(int index, double real, double imaginary) {
+        final int i = index << 1;
+        realAndImagParts[i] = real;
+        realAndImagParts[i + 1] = imaginary;
+    }
+
+    /**
+     * Replaces the element at the specified position in this list with the specified element.
+
+     * @param index - Index of the element to replace.
+     * @param element - Element to be stored at the specified position.
+     * @return the element previously at the specified position.
+     * @throws IndexOutOfBoundsException - if index isn't within the range of list.
+     */
+    @Override
+    public Complex set(int index, Complex element) {
+        rangeCheck(index);
+        final int i = index << 1;
+        final Complex oldValue = Complex.ofCartesian(realAndImagParts[i], realAndImagParts[i + 1]);
+        setNoRangeCheck(index, element.getReal(), element.getImaginary());
+        return oldValue;
+    }
+
+    /**
+     * Increases the capacity of this ComplexList instance, if necessary, to ensure that it can hold at
+     * least the number of elements specified by the minimum capacity argument.
+     *
+     * @param minCapacity – Desired minimum capacity.
+     * @throws OutOfMemoryError - if the {@code minCapacity} is greater than {@code MAX_ARRAY_SIZE}.
+     */
+    public void ensureCapacity(int minCapacity) {
+        ensureCapacityInternal(minCapacity);
+    }
+
+    /**
+     * Increases the capacity of this ComplexList instance, if necessary, to ensure that it can hold at
+     * least the number of elements specified by the minimum capacity argument.
+     *
+     * @param minCapacity – Desired minimum capacity.
+     * @return the backing double array.
+     * @throws OutOfMemoryError - if the {@code minCapacity} is greater than {@code MAX_ARRAY_SIZE}.
+     */
+    private double[] ensureCapacityInternal(int minCapacity) {
+        modCount++;
+        final long minArrayCapacity = minCapacity * 2L;
+        if (minArrayCapacity > MAX_ARRAY_SIZE) {
+            throw new OutOfMemoryError(String.format(OOM_ERROR_STRING, minArrayCapacity));
+        }
+        final long oldArrayCapacity = realAndImagParts.length;
+        if (minArrayCapacity > oldArrayCapacity) {
+            long newArrayCapacity = oldArrayCapacity + (oldArrayCapacity >> 1);
+            if (newArrayCapacity % 2 != 0) {
+                ++newArrayCapacity;
+            }
+            if (newArrayCapacity > MAX_ARRAY_SIZE) {
```

Comment:
This could do with a note explaining that minArrayCapacity has already been checked against MAX_ARRAY_SIZE so we just clip the new capacity to the appropriate range: [minArrayCapacity, MAX_ARRAY_SIZE]